### PR TITLE
Abstracted out BasicButton and BasicToolbox Forms/displays.

### PR DESCRIPTION
**Pull request by doylejd**
- Abstracted out the BasicButton and BasicToolbox forms to allow for reuse in field/block
- Abstracted out displays for BasicButton and BasicToolbox to allow for reuse in field/block
- Updated Block settings form to allow for Type selection and conditional display of subfields
- Updated Block/Fields Form and Display functions to call shared functions
- Hooked up Services validation function for the BasicToolbox form

### CODE REVIEW

**Comment by doylejd** [comment]
Resolves issue https://github.com/d8-contrib-modules/addthis/issues/20

File: addthis.module [hunk 1]
```diff
@@ -42,3 +42,26 @@ function _addthis_field_info_formatter_field_type($field_type = NULL) {
 }
 
 
+/**
+ * Validation for services for BasicToolbox.
+ * @param array $element
+ * @param FormStateInterface $form_state
+ */
+function addThisDisplayElementServicesValidate(array $element, FormStateInterface $form_state)
```
Comment (nerdstein) [inline addthis.module]: This needs to go into a controller class

File: addthis_block/src/Plugin/Block/AddThisBlock.php [hunk 2]
```diff
@@ -26,173 +25,109 @@ class AddThisBlock extends BlockBase
...
-    $this->configuration['share_services'] = $form_state->getValue(['settings', 'addthis_settings', 'share_services']);
-    $this->configuration['buttons_size'] = $form_state->getValue(['settings', 'addthis_settings', 'buttons_size']);
-    $this->configuration['counter_orientation'] = $form_state->getValue(['settings', 'addthis_settings', 'counter_orientation']);
-    $this->configuration['extra_css'] = $form_state->getValue(['settings', 'addthis_settings', 'extra_css']);
+
+  public function addthisAjaxCallback(array $form, FormStateInterface $form_state){
+    return $form['settings']['settings']['addthis_settings']['type_settings'];
+  }
+
+  public function blockSubmit($form, FormStateInterface $form_state)
```
Comment (nerdstein) [inline addthis_block/src/Plugin/Block/AddThisBlock.php]: I'd start commenting these functions or use inheritdoc if it's a standard function in the class

File: addthis_block/src/Plugin/Block/AddThisBlock.php [hunk 3]
```diff
@@ -26,173 +25,109 @@ class AddThisBlock extends BlockBase
...
+    $this->configuration['basic_button']['button_size'] = $form_state->getValue(['settings', 'addthis_settings', 'type_settings', 'basic_button', 'button_size']);
+    $this->configuration['basic_button']['extra_css'] = $form_state->getValue(['settings', 'addthis_settings', 'type_settings', 'basic_button', 'extra_css']);
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build()
+  {
```
Comment (nerdstein) [inline addthis_block/src/Plugin/Block/AddThisBlock.php]: This code formatting is not going to pass. No big deal right now, but should note this moving forward

File: addthis_fields/src/Plugin/Field/FieldFormatter/AddThisBasicButtonFormatter.php [hunk 4]
```diff
@@ -67,44 +51,11 @@ public function settingsForm(array $form, FormStateInterface $form_state) {
...
-    // Create img button.
-    $image = array(
-      '#type' => 'addthis_element',
-      '#tag' => 'img',
-      '#attributes' => array(
-        'src' => $button_img,
-        'alt' => t('Share page with AddThis'),
-      ),
+    $settings = $this->getSettings();
+    $markup = AddThis::getInstance()->getBasicButtonMarkup($settings);
```
Comment (nerdstein) [inline addthis_fields/src/Plugin/Field/FieldFormatter/AddThisBasicButtonFormatter.php:55]: Nice and clean :)

File: src/AddThis.php [hunk 5]
```diff
@@ -8,6 +8,7 @@
 
 use Drupal\addthis\Util\AddThisJson;
 use Drupal\Component\Utility\SafeMarkup;
+use Drupal\addthis\Services\AddThisScriptManager;
 
 class AddThis {
```
Comment (nerdstein) [inline src/AddThis.php:13]: This sure looks like a controller to me ;)

**Comment by nerdstein** [comment]
@doylejd - big things from my review:
1. module files should only contain hooks, all custom callbacks should be in controller classes
2. The 'AddThis' class itself looks like a potential controller.

Controllers should go into src/Controller

**Comment by doylejd** [comment]
@nerdstein Added comments and refactored code based on standards. 

Investigating Controller classes and the ability to move the validate function. Additional refactoring may be needed. If so, I will open new issues for it.

**Pull request merged**